api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -526,25 +526,25 @@ public static StructType of(List<NestedField> fields) {
       return new StructType(fields, false);
     }
 
-    public static StructType of(List<NestedField> fields, boolean isUnionSchema) {
-      return new StructType(fields, isUnionSchema);
+    public static StructType of(List<NestedField> fields, boolean convertedFromUnionSchema) {
+      return new StructType(fields, convertedFromUnionSchema);
```

Comment:
I still don't think that Iceberg needs to model that this was converted. The conversion back to a union schema is not something that Iceberg should support. Iceberg must only _write_ Iceberg data files and those are not allowed to contain Avro unions. The use case that we aim to support is reading data files with existing unions, if I understand correctly.

Reply:
there is no other way to pass this information for `GenericAvroWriter` as it needs to know that input generic record has union schema and valuewriters needs to be created accordingly. otherwise we get cast exception. this avoids creating new generic object with record schema ( which is really expensive IMO).  what alternative is possible if don't want to add `convertedFromUnionSchema`?

Reply:
> there is no other way to pass this information for GenericAvroWriter

I understand that there is no other way to pass the information. What I'm saying is that GenericAvroWriter cannot write non-option unions anyway. Iceberg must not write anything that is not allowed by the spec. This feature can be used to read existing Avro data files, but cannot be used to write Avro data files with unions.

Reply:
+ 1. We shouldn't support writing unions.

Reply:
`but cannot be used to write Avro data files with unions.` this feature is converting union schema to record similar to spark-avro and allows iceberg to ingest data even if there is union schema (similar to spark avro). there is open questions about schema evolution which needs to be addressed.
why we need this :  avro support union schema and writing such schema to iceberg fails as it is not supported. with this feature we will support writing and transformation will be taken care by iceberg client instead of performing transformation externally.

Reply:
Reading data written with a union schema is okay, but Iceberg cannot write data files with a union schema (other than options).